glean-core/csharp/Glean/Metrics/LabeledMetricType.cs

```diff
@@ -0,0 +1,153 @@
+    /// </summary>
+    public sealed class LabeledMetricType<T> where T : ILabeledSubmetricInterface
+    {
+        private bool disabled;
+        private string[] sendInPings;
+        private T submetric;
+        private UInt64 handle;
+
+        /// <exception cref="InvalidOperationException">
+        /// If the type of the `submetric` parameter is currently not supported.
+        /// </exception>
+        public LabeledMetricType(
+            bool disabled,
+            string category,
+            Lifetime lifetime,
+            string name,
+            string[] sendInPings,
+            T submetric,
+            HashSet<string> labels = null
+            )
+        {
+            this.disabled = disabled;
+            this.sendInPings = sendInPings;
+            this.submetric = submetric;
+
+            Func<string, string, string[], int, int, bool, string[], int, UInt64> metricTypeInstantiator;
+            switch (submetric)
+            {
+                case BooleanMetricType _:
+                    metricTypeInstantiator = LibGleanFFI.glean_new_labeled_boolean_metric;
+                    break;
+                case StringMetricType _:
+                    metricTypeInstantiator = LibGleanFFI.glean_new_labeled_string_metric;
+                    break;
+                default:
+                    throw new InvalidOperationException("Can not create a labeled version of this metric type");
+            }
+
+            handle = metricTypeInstantiator(
+                    category,
+                    name,
+                    sendInPings,
+                    sendInPings.Length,
+                    (int)lifetime,
+                    disabled,
+                    labels?.ToArray(),
+                    (labels != null) ? labels.Count : 0);
+        }
+
+        /// <summary>
+        /// Get the specific metric for a given label.
+        /// 
+        /// If a set of acceptable labels were specified in the metrics.yaml file,
+        /// and the given label is not in the set, it will be recorded under the
+        /// special `__other__`.
+        /// 
+        /// If a set of acceptable labels was not specified in the metrics.yaml file,
+        /// only the first 16 unique labels will be used. After that, any additional
+        /// labels will be recorded under the special `__other__` label.
+        /// 
+        /// Labels must be snake_case and less than 30 characters. If an invalid label
+        /// is used, the metric will be recorded in the special `__other__` label.
+        /// </summary>
+        /// <param name="label">The label</param>
+        /// <exception cref="InvalidOperationException">
+        /// If the type of  `T`  is currently not supported.
+        /// </exception>
+        public T this[string label]
+        {
+            get {
+                // Note the double `(T)(ILabeledSubmetricInterface)` cast before returning. This is
```

Comment:
Thanks for leaving such a detailed comment, this technique of the double cast is weird!